File: substrate/fullscreen_draft_before_merge_floating_multiselect_compose.py
"""Fullscreen-open over draft-before-merge floating multi-select model decision (pure).

live_dispatched / merge_executed / draft_written always False.
live_router_authorized / secrets_stored / remote_index_queried always False.
pdf_primary always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.draft_before_merge_floating_multiselect_model_decision_compose import (
    DraftBeforeMergeFloatingMultiselectModelDecisionCompose,
    DraftBeforeMergeFloatingMultiselectModelDecisionComposeError,
    compose_draft_before_merge_floating_multiselect_model_decision,
)
from substrate.floating_fullscreen_open_compose import (
    FloatingFullscreenOpenCompose,
    FloatingFullscreenOpenComposeError,
    compose_floating_fullscreen_open,
)


class FullscreenDraftBeforeMergeFloatingMultiselectComposeError(ValueError):
    """Fail-closed validation for fullscreen + draft multi-select model decision pack."""


@dataclass(frozen=True)
class FullscreenDraftBeforeMergeFloatingMultiselectCompose:
    session_id: str
    parent_asset_id: str
    week_id: str
    asset_id: str
    title: str
    account_id: str
    fullscreen: FloatingFullscreenOpenCompose
    draft_pack: DraftBeforeMergeFloatingMultiselectModelDecisionCompose
    session_aligned: bool
    parent_aligned: bool
    pack_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    draft_written: bool
    analysis_written: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    remote_index_queried: bool
    backlog_mutated: bool
    store_mutated: bool
    suite_rewritten: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    twin_written: bool
    prompts_injected: bool
    live_dispatch_authorized: bool
    inventory_mutated: bool
    live_execution_authorized: bool
    charge_executed: bool
    record_persisted: bool
    purchase_executed: bool
    hosted: bool
    remote_fetched: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "fullscreen": self.fullscreen.to_dict(),
            "draft_pack": self.draft_pack.to_dict(),
            "session_aligned": self.session_aligned,
            "parent_aligned": self.parent_aligned,
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "draft_written": False,
            "analysis_written": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "remote_index_queried": False,
            "backlog_mutated": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "twin_written": False,
            "prompts_injected": False,
            "live_dispatch_authorized": False,
            "inventory_mutated": False,
            "live_execution_authorized": False,
            "charge_executed": False,
            "record_persisted": False,
            "purchase_executed": False,
            "hosted": False,
            "remote_fetched": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "fullscreen_draft_before_merge_floating_multiselect_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_fullscreen_draft_before_merge_floating_multiselect(
    *,
    fullscreen: object,
    draft_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> FullscreenDraftBeforeMergeFloatingMultiselectCompose:
    """Fullscreen-open + draft-before-merge multi-select model decision. Never dispatches."""
    if not isinstance(operator_ack, bool):
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(fullscreen, dict):
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            "fullscreen must be an object"
        )
    if not isinstance(draft_pack, dict):
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            "draft_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · merge_executed=false · draft_written=false",
        "live_router_authorized=false · secrets_stored=false · remote_index_queried=false",
        "production_router_verdict=REJECT",
    ]

    try:
        fs = compose_floating_fullscreen_open(
            session_id=fullscreen.get("session_id"),
            parent_asset_id=fullscreen.get("parent_asset_id"),
            operator_ack=operator_ack,
            existing_instance=fullscreen.get("existing_instance"),
            highlight=fullscreen.get("highlight"),
            prompt=fullscreen.get("prompt"),
            gated=fullscreen.get("gated"),
            tray_siblings=fullscreen.get("tray_siblings"),
        )
    except FloatingFullscreenOpenComposeError as e:
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            str(e)
        ) from e
    notes.extend(f"[fullscreen] {n}" for n in fs.notes)

    try:
        dp = compose_draft_before_merge_floating_multiselect_model_decision(
            draft_gate=draft_pack.get("draft_gate"),
            multi_pack=draft_pack.get("multi_pack"),
            operator_ack=operator_ack,
            require_both=draft_pack.get("require_both"),
        )
    except DraftBeforeMergeFloatingMultiselectModelDecisionComposeError as e:
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            str(e)
        ) from e
    notes.extend(f"[draft_pack] {n}" for n in dp.notes)

    session = _require_nonempty(fs.session_id, field="session_id")
    parent = _require_nonempty(fs.parent_asset_id, field="parent_asset_id")
    week = _require_nonempty(dp.week_id, field="week_id")
    asset = _require_nonempty(dp.asset_id, field="asset_id")
    title = _require_nonempty(dp.title, field="title")
    account = _require_nonempty(dp.account_id, field="account_id")

    session_aligned = dp.session_id == session
    parent_aligned = (
        dp.parent_asset_id == parent or dp.asset_id == parent
    )
    if not session_aligned:
        notes.append(
            "session_id mismatch between fullscreen and draft_pack — pack_ready blocked"
        )
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between fullscreen and draft_pack — pack_ready blocked"
        )

    if require:
        pack_ready = (
            session_aligned
            and parent_aligned
            and fs.fullscreen_ready is True
            and dp.pack_ready is True
            and fs.live_dispatched is False
            and fs.merge_executed is False
            and fs.pack_dispatched is False
            and dp.draft_written is False
            and dp.merge_executed is False
            and dp.live_dispatched is False
            and dp.live_router_authorized is False
            and dp.secrets_stored is False
            and dp.remote_index_queried is False
            and dp.pdf_primary is False
            and dp.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned
            and parent_aligned
            and operator_ack is True
            and fs.merge_executed is False
            and dp.production_router_verdict == "REJECT"
            and dp.pdf_primary is False
            and (fs.fullscreen_ready is True or dp.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — fullscreen + draft-before-merge multi-select model "
            "decision ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — fullscreen, draft_pack, alignment, or operator_ack gate open"
        )

    if (
        fs.live_dispatched is not False
        or fs.merge_executed is not False
        or fs.pack_dispatched is not False
        or dp.draft_written is not False
        or dp.merge_executed is not False
        or dp.live_dispatched is not False
        or dp.live_router_authorized is not False
        or dp.secrets_stored is not False
        or dp.remote_index_queried is not False
        or dp.pdf_primary is not False
        or dp.production_router_verdict != "REJECT"
    ):
        raise FullscreenDraftBeforeMergeFloatingMultiselectComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "draft_written=false",
            "analysis_written=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "remote_index_queried=false",
            "backlog_mutated=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "twin_written=false",
            "prompts_injected=false",
            "live_dispatch_authorized=false",
            "inventory_mutated=false",
            "live_execution_authorized=false",
            "charge_executed=false",
            "record_persisted=false",
            "purchase_executed=false",
            "hosted=false",
            "remote_fetched=false",
            "production_router_verdict=REJECT",
        )
    )

    return FullscreenDraftBeforeMergeFloatingMultiselectCompose(
        session_id=session,
        parent_asset_id=parent,
        week_id=week,
        asset_id=asset,
        title=title,
        account_id=account,
        fullscreen=fs,
        draft_pack=dp,
        session_aligned=session_aligned,
        parent_aligned=parent_aligned,
        pack_ready=pack_ready,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        draft_written=False,
        analysis_written=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        remote_index_queried=False,
        backlog_mutated=False,
        store_mutated=False,
        suite_rewritten=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        twin_written=False,
        prompts_injected=False,
        live_dispatch_authorized=False,
        inventory_mutated=False,
        live_execution_authorized=False,
        charge_executed=False,
        record_persisted=False,
        purchase_executed=False,
        hosted=False,
        remote_fetched=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "fullscreen_draft_before_merge_floating_multiselect_compose_advisory"
        ),
    )


def format_fullscreen_draft_before_merge_floating_multiselect_summary(
    c: FullscreenDraftBeforeMergeFloatingMultiselectCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"fullscreen_ready={c.fullscreen.fullscreen_ready} · "
        f"draft_pack_ready={c.draft_pack.pack_ready} · "
        f"session_aligned={c.session_aligned} · "
        f"parent_aligned={c.parent_aligned} · "
        f"verdict={c.production_router_verdict} · "
        "live_dispatched=false · merge_executed=false · draft_written=false"
    )


__all__ = [
    "FullscreenDraftBeforeMergeFloatingMultiselectCompose",
    "FullscreenDraftBeforeMergeFloatingMultiselectComposeError",
    "compose_fullscreen_draft_before_merge_floating_multiselect",
    "format_fullscreen_draft_before_merge_floating_multiselect_summary",
]

File: apps/reading/src/api/fullscreenDraftBeforeMergeFloatingMultiselectCompose.ts
/**
 * Floating fullscreen-open residual over draft-before-merge floating
 * multi-select model decision pack (pure).
 *
 * Operator vision: open a floating deep-research instance fullscreen while
 * provisional draft-before-merge + multi-select cohesive unit + model decision
 * budget + twin search free settings remain pure.
 *
 * live_dispatched / merge_executed / draft_written always false.
 * live_router_authorized / secrets_stored / remote_index_queried always false.
 * production_router_verdict always REJECT.
 */

import {
  composeFloatingFullscreenOpen,
  type FloatingFullscreenOpenCompose,
  type FloatingFullscreenOpenInput,
} from "./floatingFullscreenOpenCompose";
import {
  composeDraftBeforeMergeFloatingMultiselectModelDecision,
  type DraftBeforeMergeFloatingMultiselectModelDecisionCompose,
  type DraftBeforeMergeFloatingMultiselectModelDecisionInput,
} from "./draftBeforeMergeFloatingMultiselectModelDecisionCompose";

export interface FullscreenDraftBeforeMergeFloatingMultiselectInput {
  fullscreen: Omit<FloatingFullscreenOpenInput, "operator_ack">;
  draft_pack: Omit<
    DraftBeforeMergeFloatingMultiselectModelDecisionInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface FullscreenDraftBeforeMergeFloatingMultiselectCompose {
  session_id: string;
  parent_asset_id: string;
  week_id: string;
  asset_id: string;
  title: string;
  account_id: string;
  fullscreen: FloatingFullscreenOpenCompose;
  draft_pack: DraftBeforeMergeFloatingMultiselectModelDecisionCompose;
  session_aligned: boolean;
  parent_aligned: boolean;
  pack_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  draft_written: false;
  analysis_written: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  remote_index_queried: false;
  backlog_mutated: false;
  store_mutated: false;
  suite_rewritten: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  twin_written: false;
  prompts_injected: false;
  live_dispatch_authorized: false;
  inventory_mutated: false;
  live_execution_authorized: false;
  charge_executed: false;
  record_persisted: false;
  purchase_executed: false;
  hosted: false;
  remote_fetched: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "fullscreen_draft_before_merge_floating_multiselect_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Fullscreen-open stacked on draft-before-merge floating multi-select pack.
 * Never live-dispatches; never merges; ND REJECT.
 */
export function composeFullscreenDraftBeforeMergeFloatingMultiselect(
  input: FullscreenDraftBeforeMergeFloatingMultiselectInput,
): FullscreenDraftBeforeMergeFloatingMultiselectCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.fullscreen || typeof input.fullscreen !== "object") {
    throw new Error("fullscreen must be an object");
  }
  if (!input.draft_pack || typeof input.draft_pack !== "object") {
    throw new Error("draft_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · merge_executed=false · draft_written=false",
    "live_router_authorized=false · secrets_stored=false · remote_index_queried=false",
    "production_router_verdict=REJECT",
  ];

  const fullscreen = composeFloatingFullscreenOpen({
    ...input.fullscreen,
    operator_ack: input.operator_ack,
  });
  notes.push(...fullscreen.notes.map((n) => `[fullscreen] ${n}`));

  const draft_pack = composeDraftBeforeMergeFloatingMultiselectModelDecision({
    ...input.draft_pack,
    operator_ack: input.operator_ack,
  });
  notes.push(...draft_pack.notes.map((n) => `[draft_pack] ${n}`));

  const session_id = requireNonEmpty(fullscreen.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    fullscreen.parent_asset_id,
    "parent_asset_id",
  );
  const week_id = requireNonEmpty(draft_pack.week_id, "week_id");
  const asset_id = requireNonEmpty(draft_pack.asset_id, "asset_id");
  const title = requireNonEmpty(draft_pack.title, "title");
  const account_id = requireNonEmpty(draft_pack.account_id, "account_id");

  const session_aligned = draft_pack.session_id === session_id;
  const parent_aligned =
    draft_pack.parent_asset_id === parent_asset_id ||
    draft_pack.asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between fullscreen and draft_pack — pack_ready blocked",
    );
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between fullscreen and draft_pack — pack_ready blocked",
    );
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      fullscreen.fullscreen_ready === true &&
      draft_pack.pack_ready === true &&
      fullscreen.live_dispatched === false &&
      fullscreen.merge_executed === false &&
      fullscreen.pack_dispatched === false &&
      draft_pack.draft_written === false &&
      draft_pack.merge_executed === false &&
      draft_pack.live_dispatched === false &&
      draft_pack.live_router_authorized === false &&
      draft_pack.secrets_stored === false &&
      draft_pack.remote_index_queried === false &&
      draft_pack.pdf_primary === false &&
      draft_pack.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      fullscreen.merge_executed === false &&
      draft_pack.production_router_verdict === "REJECT" &&
      draft_pack.pdf_primary === false &&
      (fullscreen.fullscreen_ready === true || draft_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — fullscreen + draft-before-merge multi-select model decision ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — fullscreen, draft_pack, alignment, or operator_ack gate open",
    );
  }

  if (
    fullscreen.live_dispatched !== false ||
    fullscreen.merge_executed !== false ||
    fullscreen.pack_dispatched !== false ||
    draft_pack.draft_written !== false ||
    draft_pack.merge_executed !== false ||
    draft_pack.live_dispatched !== false ||
    draft_pack.live_router_authorized !== false ||
    draft_pack.secrets_stored !== false ||
    draft_pack.remote_index_queried !== false ||
    draft_pack.pdf_primary !== false ||
    draft_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("remote_index_queried=false");
  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("inventory_mutated=false");
  notes.push("live_execution_authorized=false");
  notes.push("charge_executed=false");
  notes.push("record_persisted=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("remote_fetched=false");
  notes.push("production_router_verdict=REJECT");

  return {
    session_id,
    parent_asset_id,
    week_id,
    asset_id,
    title,
    account_id,
    fullscreen,
    draft_pack,
    session_aligned,
    parent_aligned,
    pack_ready,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    draft_written: false,
    analysis_written: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    remote_index_queried: false,
    backlog_mutated: false,
    store_mutated: false,
    suite_rewritten: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    twin_written: false,
    prompts_injected: false,
    live_dispatch_authorized: false,
    inventory_mutated: false,
    live_execution_authorized: false,
    charge_executed: false,
    record_persisted: false,
    purchase_executed: false,
    hosted: false,
    remote_fetched: false,
    production_router_verdict: "REJECT",
    notes,
    authority:
      "fullscreen_draft_before_merge_floating_multiselect_compose_advisory",
  };
}

export function formatFullscreenDraftBeforeMergeFloatingMultiselectSummary(
  c: FullscreenDraftBeforeMergeFloatingMultiselectCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `fullscreen_ready=${c.fullscreen.fullscreen_ready} · ` +
    `draft_pack_ready=${c.draft_pack.pack_ready} · ` +
    `session_aligned=${c.session_aligned} · ` +
    `parent_aligned=${c.parent_aligned} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_dispatched=false · merge_executed=false · draft_written=false`
  );
}
